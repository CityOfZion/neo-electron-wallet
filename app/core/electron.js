import { shell } from 'electron'

export const openExternal = (srcLink) => {
  shell.openExternal(srcLink)
}
