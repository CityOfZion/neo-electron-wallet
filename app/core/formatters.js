// @flow
import { ASSETS } from './constants'
import numeral from 'numeral'

import { truncateNumber } from './math'
import numeral from 'numeral'

export const COIN_DECIMAL_LENGTH = 8
export const SHORT_DISPLAY_DECIMAL_LENGTH = 4

const formatWithDecimals = (value: number | string, decimals: number) => numeral(value).format(`0,0.${'0'.repeat(decimals)}`)

export const formatGAS = (value: number | string, shortDisplay: boolean = false): string => {
  const decimals = shortDisplay ? SHORT_DISPLAY_DECIMAL_LENGTH : COIN_DECIMAL_LENGTH
  const result = formatWithDecimals(truncateNumber(value, decimals), decimals)
  // https://github.com/adamwdraper/Numeral-js/issues/575
  if (result === 'NaN') {
    return toFixedDecimals(value, decimals)
  }
  return result
}

export const formatThousands = (value: number | string): string => numeral(value).format('0,0')

export const formatNEO = (value: number | string): string => formatThousands(value)

export const formatBalance = (symbol: SymbolType, balance: number, shortDisplay: boolean = false) => {
  if (symbol === ASSETS.NEO) {
    return formatNEO(balance)
  }
  return formatGAS(balance, shortDisplay)
}

export const toFixedDecimals = (value: number | string, decimals: number) => truncateNumber(parseFloat(value), decimals).toFixed(decimals)

export const formatFiat = (value: number | string): string => numeral(value).format('0,0.00')
