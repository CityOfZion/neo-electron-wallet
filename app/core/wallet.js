// @flow
import { ASSETS, TOKENS } from './constants'
import { getAccountFromWIFKey, verifyAddress } from 'neon-js'

const MIN_PASSPHRASE_LEN = 4

export const validatePassphrase = (passphrase: string): boolean => passphrase.length >= MIN_PASSPHRASE_LEN

export const checkMatchingPassphrases = (passphrase: string, passphrase2: string) => passphrase !== passphrase2

export const verifyPrivateKey = (wif: string): boolean => {
  if (!wif) {
    return false
  }
  const account = getAccountFromWIFKey(wif)
  return account !== -1 && account.address
}

export const isToken = (symbol: TokenSymbol) => Object.keys(TOKENS).includes(symbol)

export const obtainTokenBalance = (tokens: Array<TokenType>, symbol: TokenSymbol) => {
  if (!isToken(symbol)) {
    throw new Error(`${symbol} is not a valid token`)
  }
  return tokens.find((token: TokenType) => token.symbol === symbol)
}

export const validateTransactionBeforeSending = (neoBalance: number, gasBalance: number, tokenBalance: number, symbol: TokenSymbol, sendAddress: string, sendAmount: string) => {
  if (!sendAddress || !sendAmount) {
    return {
      error: 'Please specify an address and amount',
      valid: false
    }
  }

  if (symbol !== ASSETS.NEO && symbol !== ASSETS.GAS && !isToken(symbol)) {
    return {
      error: 'That asset is not NEO, GAS or NEP-5 Token',
      valid: false
    }
  }

  try {
    if (verifyAddress(sendAddress) !== true || sendAddress.charAt(0) !== 'A') {
      return {
        error: 'The address you entered was not valid.',
        valid: false
      }
    }
  } catch (e) {
    return {
      error: 'The address you entered was not valid.',
      valid: false
    }
  }

  if (symbol === ASSETS.NEO) {
    if (parseFloat(sendAmount) !== parseInt(sendAmount)) { // check for fractional neo
      return {
        error: 'You cannot send fractional amounts of NEO.',
        valid: false
      }
    }
    if (parseInt(sendAmount) > neoBalance) { // check for value greater than account balance
      return {
        error: 'You do not have enough NEO to send.',
        valid: false
      }
    }
  } else if (symbol === ASSETS.GAS) {
    if (parseFloat(sendAmount) > gasBalance) {
      return {
        error: 'You do not have enough GAS to send.',
        valid: false
      }
    }
  } else {
    if (parseFloat(sendAmount) > tokenBalance) {
      return {
        error: `You do not have enough ${symbol} to send.`,
        valid: false
      }
    }
  }

  if (parseFloat(sendAmount) < 0) { // check for negative asset
    return {
      error: 'You cannot send negative amounts of an asset.',
      valid: false
    }
  }

  return {
    error: '',
    valid: true
  }
}
