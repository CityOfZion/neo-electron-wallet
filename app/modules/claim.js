// @flow
import { api, rpc } from 'neon-js'

import {
  showErrorNotification,
  showSuccessNotification,
  showInfoNotification
} from './notifications'
import {
  getNetwork,
  getWIF,
  getAddress,
  getSigningFunction,
  getPublicKey,
  getIsHardwareLogin,
  getNEO
} from '../core/deprecated'
import { ASSETS } from '../core/constants'
import { FIVE_MINUTES_MS } from '../core/time'

import { log } from '../util/Logs'

// Constants
export const SET_CLAIM_REQUEST = 'SET_CLAIM_REQUEST'
export const DISABLE_CLAIM = 'DISABLE_CLAIM'
export const SET_FINALIZE_CLAIM = 'SET_FINALIZE_CLAIM'
export const SET_TRANSACTION_ID = 'SET_TRANSACTION_ID'

// Actions
export function setClaimRequest (claimRequest: boolean) {
  return {
    type: SET_CLAIM_REQUEST,
    payload: { claimRequest }
  }
}
export function setTransactionId (transactionId: string) {
  return {
    type: SET_TRANSACTION_ID,
    payload: { transactionId }
  }
}

export function setFinalizeClaim (finalizeClaim: boolean) {
  return {
    type: SET_FINALIZE_CLAIM,
    payload: { finalizeClaim }
  }
}

export function disableClaim (disableClaimButton: boolean) {
  return {
    type: DISABLE_CLAIM,
    payload: { disableClaimButton }
  }
}

export const doClaimNotify = () => async (
  dispatch: DispatchType,
  getState: GetStateType
) => {
  const state = getState()
  const address = getAddress(state)
  const net = getNetwork(state)
  const signingFunction = getSigningFunction(state)
  const publicKey = getPublicKey(state)
  const privateKey = getWIF(state)
  const isHardwareClaim = getIsHardwareLogin(state)

  log(net, 'CLAIM', address, { info: 'claim all GAS' })

  if (isHardwareClaim) {
    dispatch(showInfoNotification({
      message: 'Sign transaction 2 of 2 to claim GAS on your hardware device (claiming GAS)',
      autoDismiss: 0
    }))
  }

  try {
    const { response } = await api.claimGas({ net, address, publicKey, privateKey, signingFunction })

    if (!response.result) {
      throw new Error('Claim failed')
    }

    dispatch(showSuccessNotification({
      message: 'Claim was successful! Your balance will update once the blockchain has processed it.'
    }))
    setTimeout(() => dispatch(disableClaim(false)), FIVE_MINUTES_MS)
  } catch (err) {
    return dispatch(showErrorNotification({ message: 'Claim failed' }))
  }
}

// To initiate claim, first send all NEO to own address, the set claimRequest state
// When new claims are available, this will trigger the claim
export const doGasClaim = () => async (
  dispatch: DispatchType,
  getState: GetStateType
) => {
  const state = getState()
  const address = getAddress(state)
  const net = getNetwork(state)
  const NEO = getNEO(state)
  const publicKey = getPublicKey(state)
  const privateKey = getWIF(state)
  const signingFunction = getSigningFunction(state)
  const isHardwareClaim = getIsHardwareLogin(state)

  // if no NEO in account, no need to send to self first
  if (NEO === '0') {
    return dispatch(doClaimNotify())
  } else {
    dispatch(showInfoNotification({
      message: 'Sending NEO to Yourself...',
      autoDismiss: 0
    }))
    log(net, 'SEND', address, { to: address, amount: NEO, asset: ASSETS.NEO })

    if (isHardwareClaim) {
      dispatch(showInfoNotification({
        message: 'Sign transaction 1 of 2 to claim GAS on your hardware device (sending NEO to yourself)',
        autoDismiss: 0
      }))
    }

    try {
      const { response } = await api.sendAsset({
        net,
        address,
        publicKey,
        privateKey,
        signingFunction,
        intents: api.makeIntent({ [ASSETS.NEO]: NEO }, address)
      })

      if (!response.result || !response.txid) {
        throw new Error('Transaction failed!')
      }

      dispatch(showInfoNotification({
        message: 'Waiting for transaction to clear...',
        autoDismiss: 0
      }))
      dispatch(setTransactionId(response.txid))
      dispatch(setClaimRequest(true))
      return dispatch(disableClaim(true))
    } catch (err) {
      return dispatch(showErrorNotification({ message: 'Transaction failed!' }))
    }
  }
}

export const checkClaimStatus = () => async (
  dispatch: DispatchType,
  getState: GetStateType
) => {
  const state = getState()
  const net = getNetwork(state)
  const txid = getTransactionId(state)
  const endpoint = await api.loadBalance(api.getRPCEndpointFrom, { net })
  const response = await rpc.Query.getRawTransaction(txid).execute(endpoint)
  if (response.result && response.result.blockhash && response.result.confirmations) {
    dispatch(setTransactionId(''))
    dispatch(setFinalizeClaim(true))
  }
}

// State Getters
export const getFinalizeClaim = (state: Object) => state.claim.finalizeClaim
export const getClaimRequest = (state: Object) => state.claim.claimRequest
export const getDisableClaimButton = (state: Object) => state.claim.disableClaimButton
export const getTransactionId = (state: Object) => state.claim.transactionId

const initialState = {
  claimRequest: false,
  disableClaimButton: false,
  transactionId: '',
  finalizeClaim: false
}

export default (state: Object = initialState, action: ReduxAction) => {
  switch (action.type) {
    case SET_TRANSACTION_ID:
      const { transactionId } = action.payload
      return {
        ...state,
        transactionId
      }
    case SET_FINALIZE_CLAIM:
      const { finalizeClaim } = action.payload
      return {
        ...state,
        finalizeClaim
      }
    case SET_CLAIM_REQUEST:
      const { claimRequest } = action.payload
      return {
        ...state,
        claimRequest
      }
    case DISABLE_CLAIM:
      const { disableClaimButton } = action.payload
      return {
        ...state,
        disableClaimButton
      }
    default:
      return state
  }
}
