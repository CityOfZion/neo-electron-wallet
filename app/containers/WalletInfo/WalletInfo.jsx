// @flow
import React, { Component } from 'react'
import { isNil } from 'lodash'
import Claim from '../../components/Claim.js'
import MdSync from 'react-icons/lib/md/sync'
import QRCode from 'qrcode/lib/browser'
import { clipboard } from 'electron'
import Copy from 'react-icons/lib/md/content-copy'
import ReactTooltip from 'react-tooltip'
import { formatGAS } from '../../core/formatters'
import { ASSETS } from '../../core/constants'

type Props = {
  address: string,
  neo: number,
  net: NetworkType,
  gas: number,
  price: number,
  initiateGetBalance: Function,
  sendEvent: Function,
  clearTransactionEvent: Function
}

export default class WalletInfo extends Component<Props> {
  canvas: ?HTMLCanvasElement

  componentDidMount () {
    const { initiateGetBalance, net, address } = this.props
    initiateGetBalance(net, address)
    QRCode.toCanvas(this.canvas, address, { version: 5 }, (err) => {
      if (err) console.log(err)
    })
  }

  // force sync with balance data
  refreshBalance = () => {
    const { sendEvent, initiateGetBalance, clearTransactionEvent, net, address } = this.props
    sendEvent(true, 'Refreshing...')
    initiateGetBalance(net, address).then((response) => {
      sendEvent(true, 'Received latest blockchain information.')
      setTimeout(() => clearTransactionEvent(), 1000)
    })
  }

  render () {
    const { address, neo, gas, price } = this.props
    if (isNil(address)) {
      return null
    }
    return (
      <div id='accountInfo'>
        <div className='label'>Your Public Neo Address:</div>
        <div className='address'>
          {address}
          <span className='copyKey' onClick={() => clipboard.writeText(address)}><Copy data-tip data-for='copyAddressTip' /></span>
        </div>
        <ReactTooltip class='solidTip' id='copyAddressTip' place='bottom' type='dark' effect='solid'>
          <span>Copy Public Address</span>
        </ReactTooltip>
        <div className='spacer' />
        <div id='balance'>
          <div className='split'>
            <div className='label'>{ASSETS.NEO}</div>
            <div className='amountBig amountNeo'>{neo}</div>
          </div>
          <div className='split'>
            <div className='label'>{ASSETS.GAS}</div>
            <div className='amountBig amountGas'>{formatGAS(gas)}</div>
          </div>
          <div className='refreshBalance' onClick={this.refreshBalance} >
            <MdSync id='refresh' data-tip data-for='refreshBalanceTip' />
            <ReactTooltip class='solidTip' id='refreshBalanceTip' place='bottom' type='dark' effect='solid'>
              <span>Refresh account balance</span>
            </ReactTooltip>
          </div>

          <div className='fiat'>US {price}</div>
        </div>
        <div className='spacer' />
        <Claim />
        <div className='spacer' />
        <div className='qrCode'><canvas id='qrCanvas' ref={(node) => { this.canvas = node }} /></div>
      </div>
    )
  }
}
