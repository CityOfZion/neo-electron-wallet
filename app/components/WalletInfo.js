import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import { isNil } from 'lodash'
import Claim from './Claim.js'
import MdSync from 'react-icons/lib/md/sync'
import QRCode from 'qrcode'
import { initiateGetBalance } from '../components/NetworkSwitch'
import { sendEvent, clearTransactionEvent } from '../modules/transactions'
import { clipboard } from 'electron'
import Copy from 'react-icons/lib/md/content-copy'
import ReactTooltip from 'react-tooltip'

// force sync with balance data
const refreshBalance = (dispatch, net, address) => {
  dispatch(sendEvent(true, 'Refreshing...'))
  initiateGetBalance(dispatch, net, address).then((response) => {
    dispatch(sendEvent(true, 'Received latest blockchain information.'))
    setTimeout(() => dispatch(clearTransactionEvent()), 1000)
  })
}

let WalletInfo = class WalletInfo extends Component {
  componentDidMount () {
    const { dispatch, net, address } = this.props
    initiateGetBalance(dispatch, net, address)
    QRCode.toCanvas(this.canvas, address, { version: 5 }, (err) => {
      if (err) console.log(err)
    })
  }

  render () {
    const { address, neo, net, dispatch, price } = this.props
    if (isNil(address)) {
      return null
    }
    return (
      <div id='accountInfo'>
        <div className='label'>Your Public Neo Address:</div>
        <div className='address'>
          {address}
          <span className='copyKey' onClick={() => clipboard.writeText(address)}><Copy data-tip data-for='copyAddressTip' /></span>
        </div>
        <ReactTooltip class='solidTip' id='copyAddressTip' place='bottom' type='dark' effect='solid'>
          <span>Copy Public Address</span>
        </ReactTooltip>
        <div className='spacer' />
        <div id='balance'>
          <div className='split'>
            <div className='label'>NEO</div>
            <div className='amountBig amountNeo'>{neo}</div>
          </div>
          <div className='split'>
            <div className='label'>GAS</div>
            <div className='amountBig amountGas'>{this.props.gas}</div>
          </div>
          <div className='refreshBalance' onClick={() => refreshBalance(dispatch, net, address)} >
            <MdSync id='refresh' data-tip data-for='refreshBalanceTip' />
            <ReactTooltip class='solidTip' id='refreshBalanceTip' place='bottom' type='dark' effect='solid'>
              <span>Refresh account balance</span>
            </ReactTooltip>
          </div>

          <div className='fiat'>US {price}</div>
        </div>
        <div className='spacer' />
        <Claim />
        <div className='spacer' />
        <div className='qrCode'><canvas id='qrCanvas' ref={(node) => { this.canvas = node }} /></div>
      </div>
    )
  }
}

const mapStateToProps = (state) => ({
  neo: state.wallet.Neo,
  gas: state.wallet.Gas,
  address: state.account.address,
  net: state.metadata.network,
  price: state.wallet.price
})

WalletInfo.propTypes = {
  dispatch: PropTypes.func.isRequired,
  address: PropTypes.string,
  neo: PropTypes.number,
  net: PropTypes.string,
  gas: PropTypes.string,
  price: PropTypes.string
}

WalletInfo = connect(mapStateToProps)(WalletInfo)

export default WalletInfo
