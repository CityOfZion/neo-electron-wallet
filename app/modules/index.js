// @flow
import { combineReducers } from 'redux'
import api from './api'
import generateWallet from './generateWallet'
import transactions from './transactions'
import wallet from './wallet'
import claim from './claim'
import notifications from './notifications'
import modal from './modal'
import addressBook from './addressBook'

export default combineReducers({
  api,
  generateWallet,
  wallet,
  transactions,
  claim,
  notifications,
  modal,
  addressBook
})
