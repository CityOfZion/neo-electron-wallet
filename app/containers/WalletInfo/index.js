// @flow
import { connect, type MapStateToProps } from 'react-redux'
import { bindActionCreators } from 'redux'
import { compose } from 'recompose'

import pricesActions from '../../actions/pricesActions'
import withData from '../../hocs/api/withData'
import withActions from '../../hocs/api/withActions'
import withNetworkData from '../../hocs/withNetworkData'
import withAccountData from '../../hocs/withAccountData'
import withCurrencyData from '../../hocs/withCurrencyData'
import { updateSettingsActions } from '../../actions/settingsActions'
import { getNetworks } from '../../core/networks'
import { showErrorNotification, showSuccessNotification } from '../../modules/notifications'
import { loadWalletData, getNEO, getGAS, getTokenBalances } from '../../modules/wallet'
import { showModal } from '../../modules/modal'
import { participateInSale, oldParticipateInSale } from '../../modules/sale'

import WalletInfo from './WalletInfo'

const mapStateToProps: MapStateToProps<*, *, *> = (state: Object) => ({
  NEO: getNEO(state),
  GAS: getGAS(state),
  tokenBalances: getTokenBalances(state),
  networks: getNetworks()
})

const mapPricesDataToProps = ({ NEO, GAS }) => ({
  neoPrice: NEO,
  gasPrice: GAS
})

const actionCreators = {
  loadWalletData,
  showErrorNotification,
  showSuccessNotification,
  showModal,
  participateInSale,
  oldParticipateInSale
}

const mapDispatchToProps = dispatch => bindActionCreators(actionCreators, dispatch)

const mapActionsToProps = (actions) => ({
  setUserGeneratedTokens: (tokens) => actions.request({ tokens })
})

export default compose(
  connect(mapStateToProps, mapDispatchToProps),
  withData(pricesActions, mapPricesDataToProps),
  withNetworkData(),
  withAccountData(),
  withCurrencyData('currencyCode'),
  withActions(updateSettingsActions, mapActionsToProps)
)(WalletInfo)
