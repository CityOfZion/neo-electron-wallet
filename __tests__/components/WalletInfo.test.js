import React from 'react'
import { Provider } from 'react-redux'
import configureStore from 'redux-mock-store'
import thunk from 'redux-thunk'
import { mount, shallow } from 'enzyme'
import { SET_TRANSACTION_HISTORY, SET_BALANCE, SET_GAS_PRICE, SET_NEO_PRICE } from '../../app/modules/wallet'
import { SHOW_NOTIFICATION, HIDE_NOTIFICATION } from '../../app/modules/notification'
import { SET_HEIGHT } from '../../app/modules/metadata'
import { SET_CLAIM } from '../../app/modules/claim'
import WalletInfo from '../../app/containers/WalletInfo'

// TODO research how to move the axios mock code which is repeated in NetworkSwitch to a helper or config file
import axios from 'axios'
import MockAdapter from 'axios-mock-adapter'
import { clipboard } from 'electron'
import { version } from '../../package.json'
import { formatFiat } from '../../app/core/formatters'

const axiosMock = new MockAdapter(axios)
axiosMock
  .onGet('http://testnet-api.wallet.cityofzion.io/v2/version')
  .reply(200, { version })
axiosMock
  .onGet('https://api.coinmarketcap.com/v1/ticker/neo/?convert=USD')
  .reply(200, [ { price_usd: 24.50 } ])
axiosMock
  .onGet('https://api.coinmarketcap.com/v1/ticker/gas/?convert=USD')
  .reply(200, [ { price_usd: 18.20 } ])

jest.mock('electron', () => ({
  clipboard: {
    writeText: jest.fn()
  }
}))
jest.mock('neon-js')

jest.unmock('qrcode')
import QRCode from 'qrcode/lib/browser' // eslint-disable-line
QRCode.toCanvas = jest.fn()

const initialState = {
  account: {
    address: 'ANqUrhv99rwCiFTL6N1An9NH5UVkPYxTuw'
  },
  metadata: {
    network: 'TestNet'
  },
  wallet: {
    Neo: 10,
    Gas: 1.0001001,
    neoPrice: 25.48,
    gasPrice: 18.10
  },
  claim: {
    claimAmount: 0.5
  }
}

const setup = (state = initialState, shallowRender = true) => {
  const store = configureStore([thunk])(state)

  let wrapper
  if (shallowRender) {
    wrapper = shallow(<WalletInfo store={store} />)
  } else {
    wrapper = mount(
      <Provider store={store}>
        <WalletInfo />
      </Provider>
    )
  }

  return {
    store,
    wrapper
  }
}

describe('WalletInfo', () => {
  test('renders without crashing', (done) => {
    const { wrapper } = setup()
    expect(wrapper).toMatchSnapshot()
    done()
  })
  test('correctly renders data from state', (done) => {
    const { wrapper } = setup(initialState, false)

    const addressField = wrapper.find('.address')
    const neoWalletValue = wrapper.find('.neoWalletValue')
    const gasWalletValue = wrapper.find('.gasWalletValue')
    const walletValue = wrapper.find('.walletTotal')
    const expectedNeoWalletValue = formatFiat(initialState.wallet.neoPrice * initialState.wallet.Neo)
    const expectedGasWalletValue = formatFiat(initialState.wallet.gasPrice * initialState.wallet.Gas)
    const expectedWalletValue = formatFiat(initialState.wallet.neoPrice * initialState.wallet.Neo + initialState.wallet.gasPrice * initialState.wallet.Gas)
    const neoField = wrapper.find('.amountNeo')
    const gasField = wrapper.find('.amountGas')

    expect(neoWalletValue.text()).toEqual(`US $${expectedNeoWalletValue}`)
    expect(gasWalletValue.text()).toEqual(`US $${expectedGasWalletValue}`)
    expect(walletValue.text()).toEqual(`Total US $${expectedWalletValue}`)
    expect(addressField.text().split('<')[0]).toEqual(initialState.account.address)
    expect(neoField.text()).toEqual(`${initialState.wallet.Neo}`)
    expect(gasField.text()).toEqual(`${initialState.wallet.Gas}`)
    done()
  })
  test('copy to clipboard is getting called on click', (done) => {
    const { wrapper } = setup()
    const deepWrapper = wrapper.dive()

    expect(clipboard.writeText.mock.calls.length).toBe(0)
    deepWrapper.find('.copyKey').simulate('click')
    expect(clipboard.writeText.mock.calls.length).toBe(1)
    done()
  })
  test('refreshBalance is getting called on click', (done) => {
    const { wrapper, store } = setup()
    const deepWrapper = wrapper.dive()

    const actionTypes = [
      SHOW_NOTIFICATION,
      SET_TRANSACTION_HISTORY,
      SET_HEIGHT,
      SET_NEO_PRICE,
      SET_GAS_PRICE,
      SET_BALANCE,
      HIDE_NOTIFICATION,
      SET_CLAIM
    ]
    deepWrapper.find('.refreshBalance').simulate('click')
    setTimeout(() => {
      const actions = store.getActions()
      expect(actions.length === 15).toEqual(true)
      actions.forEach(action => {
        expect(actionTypes.indexOf(action.type) > -1).toEqual(true)
      })
      done()
    }, 1050)
  })
  test('calls the correct number of actions after mounting', (done) => {
    const { store } = setup(initialState, false)
    const actionTypes = [
      SET_TRANSACTION_HISTORY,
      SET_HEIGHT,
      SET_CLAIM,
      SET_BALANCE,
      SET_NEO_PRICE,
      SET_GAS_PRICE
    ]

    setTimeout(() => {
      const actions = store.getActions()
      expect(actions.length).toEqual(6)
      actions.forEach(action => {
        expect(actionTypes.indexOf(action.type) > -1).toEqual(true)
      })
      done()
    }, 1050)
  })
})
