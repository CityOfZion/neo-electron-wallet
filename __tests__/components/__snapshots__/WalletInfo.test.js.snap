// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WalletInfo renders without crashing 1`] = `
<Connect(withData(Connect(withData(Connect(withData(Connect(withData(Connect(withData(withProps(Connect(withData(Connect(withActions(Connect(Connect(withProgress(LoadedNotifier))))))))))))))))))
  participateInSale={[Function]}
  showModal={[Function]}
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
/>
`;
